extension-glob-content.md

```diff
@@ -0,0 +1,159 @@
+* For testing and reproducibility, it is helpful to isolate the inputs to the language server. This is easier to do if you can guarantee that the language server will not access the file system.
+
+## Protocol
+
+Note that unlike other requests in LSP, these requests are sent by the language server to the client, not vice versa. The client, not the language server, is assumed to have full access to the workspace's contents.
+
+The language server is allowed to request file paths outside of the workspace root. (This is common for, e.g., system dependencies.) The client may choose whether or not to satisfy these requests.
+
+### Initialization
+
+`ClientCapabilities` may contain two new fields to indicate client-side support for this extension:
+
+```typescript
+interface ClientCapabilities {
+	/* ... all fields from the base ClientCapabilities ... */
+
+	/**
+	 * The client provides support for workspace/xglob.
+	 */
+	globProvider?: boolean;
+	/**
+	 * The client provides support for textDocument/content.
+	 */
+	contentProvider?: boolean;
+}
+```
+
+### Content Request
+
+The content request is sent from the server to the client to request the current content of any text document. This allows language servers to operate without accessing the file system directly.
+
+_Request_:
+* method: 'textDocument/xcontent'
+* params: `ContentParams` defined as follows:
+
+```typescript
+interface ContentParams {
+	/**
+	 * The text document to receive the content for.
+	 */
+	textDocument: TextDocumentIdentifier;
+}
+```
+
+_Response_:
+* result: `Content` defined as follows
+* error: code and message set in case an exception occurs
+
+```typescript
+interface Content {
+	/**
+	 * The text of the document.
+	 */
+	text: string;
+}
+```
+
+### Glob Request
+
+The glob request is sent from the server to the client to request a list of files in the workspace that match a glob pattern. The glob pattern must be matched only against the path (not other fragments of the URI) and must be relative to the workspace's `rootPath` (set during initialization).
+
+The glob patterns must be interpreted according to the following rules:
+* `*` matches any sequence of non-path separator characters
+* `**` when alone in a path component, matches all files and zero or more directories and subdirectories ("globstar"); does not walk symlinks
+* `![]{}?` characters are interpreted literally and do not have any special meaning (TEMP NOTE: this is to simplify the implementation of globbing; we will need to write our own simple Go globbing library, and a brief survey of many languages' globbing libraries shows wide disparities in behavior)
```

Comment:
Can we change this so that a client _may_ give these special meaning? Otherwise it will be a breaking change to add this feature in the future